segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/CacheLayout.java

```diff
@@ -0,0 +1,347 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+/**
+ * Base class that defines the memory layout for a block-based {@link CacheStorage}.
+ */
+abstract class CacheLayout {
+    //region Members
+
+    /**
+     * The maximum number of bytes that can be stored in a {@link CacheStorage} using this type of layout.
+     */
+    static final long MAX_TOTAL_SIZE = 64 * 1024 * 1024 * 1024L;
+    /**
+     * The maximum size (in bytes) of any {@link CacheStorage} entry.
+     */
+    static final int MAX_ENTRY_SIZE = 0x03FF_FFFF; // 28 bits = 256MB
+    /**
+     * Null (inexistent) address.
+     */
+    static final int NO_ADDRESS = 0; // Valid addresses have cannot be 0 since Block 0 is reserved.
+    /**
+     * Null (inexistent) Buffer-Block id.
+     */
+    static final int NO_BLOCK_ID = 0; // 0 is the same as Metadata Block Id, so it's OK to use it.
+    private final int maxBufferCount;
+    private final int blocksPerBuffer;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link CacheLayout} class and performs any necessary sanity checks.
+     */
+    CacheLayout() {
+        Preconditions.checkState(MAX_TOTAL_SIZE % bufferSize() == 0,
```

Comment:
This could cause a problem. This is an abstract class but it is calling `bufferSize()` in its constructor which is an abstract method. Classes that extend this one are **Required** to call `super()` as the first line of their constructor, before any member variables are assigned. So if a subclass relies on setting bufferSize by a value passed in at construction time, it will not be set yet, so the value here could be -1.

Reply:
That is a valid concern and I agree with that. However I couldn't think of any other way to do this. I wanted `CacheLayout` to define the structure of any layout and leave it to a derived class to actually define proper values. However I do want some basic validation that the values the derived class is providing won't cause issues. 

While in general I want to stay away from such patterns, in this case I thin it's OK since derived classes will be close to this one and any issues will be quickly spotted by running a unit test trying to instantiate this.

Reply:
Good idea. I made that change.